### PR TITLE
use ballerina dockerfile, update sh scripts, add json convert bal script

### PR DESCRIPTION
**Pull request by vordimous**
- The dockerfile uses the Ballerina image base
- run scripts add files to the necessary dir and point to the test_results.json file instead of the stdout
- a ballerina script is used to convert to the necessary output.

### CODE REVIEW

File: bin/run-in-docker.sh [hunk 1]
```diff
@@ -33,14 +33,13 @@ output_dir=$(realpath "${3%/}")
...
 
 # Build the Docker image
-docker build --rm -t exercism/test-runner .
+docker build --rm  -t exercism/ballerina-test-runner .
 
 # Run the Docker image using the settings mimicking the production environment
 docker run \
     --rm \
-    --network none \
-    --read-only \
```
Comment (ErikSchierboom) [inline bin/run-in-docker.sh]: Why did you remove these flags? The second one we don't yet enforce, but the first is a hard requirement. Could you re-add at least the first one and make sure the test runner still works?

File: bin/run-tests-in-docker.sh [hunk 2]
```diff
@@ -16,16 +16,15 @@
...
 # Build the Docker image
-docker build --rm -t exercism/test-runner .
+docker build --rm -t exercism/ballerina-test-runner .
 
 # Run the Docker image using the settings mimicking the production environment
 docker run \
     --rm \
-    --network none \
-    --read-only \
+    --network host \
```
Comment (ErikSchierboom) [inline bin/run-tests-in-docker.sh]: Same remark as above

File: bin/run.sh [hunk 3]
```diff
@@ -22,39 +22,64 @@ if [ -z "$1" ] || [ -z "$2" ] || [ -z "$3" ]; then
...
+        # In some cases, the test output might be overly verbose, in which case stripping
+        # the unneeded information can be very helpful to the student
+        # sanitized_test_output=$(printf "${test_output}" | sed -n '/Test results:/,$p')
+
+        # OPTIONAL: Manually add colors to the output to help scanning the output for errors
+        # If the test output does not contain colors to help identify failing (or passing)
+        # tests, it can be helpful to manually add colors to the output
+        # colorized_test_output=$(echo "${test_output}" \
+        #      | GREP_COLOR='01;31' grep --color=always -E -e '^(ERROR:.*|.*failed)$|$' \
+        #      | GREP_COLOR='01;32' grep --color=always -E -e '^.*passed$|$')
```
Comment (ErikSchierboom) [inline bin/run.sh]: Feel free to remove this if you're not using it

File: tests/example-syntax-error/expected_results.json [hunk 4]
```diff
@@ -1,5 +1,5 @@
 {
-  "version": 1,
+  "message": "Compiling source\n\tballerina_exercism/hello_world:0.1.0",
```
Comment (ErikSchierboom) [inline tests/example-syntax-error/expected_results.json]: This error message seems to not be very useful to the student to figure out what went wrong, Could you output more information?

File: Dockerfile [hunk 5]
```diff
@@ -1,8 +1,16 @@
...
 # install packages required to run the tests
+USER root
 RUN apk add --no-cache jq coreutils
 
-WORKDIR /opt/test-runner
-COPY . .
-ENTRYPOINT ["/opt/test-runner/bin/run.sh"]
+ENV wd /opt/test-runner
+RUN mkdir -p ${wd}/bin
+WORKDIR ${wd}
```
Comment (ErikSchierboom) [inline Dockerfile]: I don't think you need to create the `bin` directory

```suggestion
WORKDIR /opt/test-runner
```

**Review by ErikSchierboom — COMMENTED** [review]
Lovely work! I've left a couple of comments, of which the compiler error test case and the `network=none` comments are key.

File: Dockerfile (hunk 5 above)
Comment (vordimous) [reply Dockerfile]: Sounds good. I found this sample in the rust-test-runner.

File: tests/example-syntax-error/expected_results.json (hunk 4 above)
Comment (vordimous) [reply tests/example-syntax-error/expected_results.json]: I will see if there is more to the stdout here.

File: bin/run.sh (hunk 3 above)
Comment (vordimous) [reply bin/run.sh]: +1

File: bin/run-in-docker.sh (hunk 1 above)
Comment (vordimous) [reply bin/run-in-docker.sh]: Ballerina pulls in dependencies and was failing when the network mode was set to `none`. Now that everything is working Ill try to work out a builder phase that pulls dependencies.

Comment (ErikSchierboom) [reply bin/run-in-docker.sh]: Excellent!

Comment (vordimous) [reply bin/run-in-docker.sh]: I think I have a working solution for the network issue. But the `--read-only` flag prevents the test script from writing the results.json file. Will it be ok to remove the read-only requirement?

Comment (vordimous) [reply bin/run-in-docker.sh]: Ok, after many tries, I think I got it working with a multi-stage build and lib caching.

**Comment by vordimous** [comment]
@ErikSchierboom, after many tries, I think I got it working with a multi-stage build and lib caching.

File: bin/run-in-docker.sh (hunk 1 above)
Comment (ErikSchierboom) [reply bin/run-in-docker.sh]: > But the --read-only flag prevents the test script from writing the results.json file. Will it be ok to remove the read-only requirement?

You should be able to write the results.json to the output directory, which is the third argument that gets passed in.

File: bin/run.sh [hunk 6]
```diff
@@ -20,41 +20,57 @@ if [ -z "$1" ] || [ -z "$2" ] || [ -z "$3" ]; then
...
-test_output=$(false)
-# TODO: substitute "false" with the actual command to run the test:
-# test_output=$(command_to_run_tests 2>&1)
+# Run the tests for the provided implementation file
+cd "$solution_dir" || exit
+# Output is saved at this location
+echo "==== ${slug}: testing..."
+test_output=""
+if ! [ -s *.bal ]; then
+    test_output="WARNING: student did not add code to the balllerina file."
```
Comment (ErikSchierboom) [inline bin/run.sh:39]: Isn't this an error?

File: bin/run.sh [hunk 7]
```diff
@@ -20,41 +20,57 @@ if [ -z "$1" ] || [ -z "$2" ] || [ -z "$3" ]; then
...
+fi
+if [ ! -e Ballerina.toml ]; then
+    test_output="$test_output \n WARNING: student did not upload Ballerina.toml."
+fi
+if [ ! -e Dependencies.toml ]; then
+    test_output="$test_output \n WARNING: student did not upload Dependencies.toml."
+fi
```
Comment (ErikSchierboom) [inline bin/run.sh:46]: Note that these conditions should never match.
What the test runner receives is not just what the student submitted, but a "merge" of the exercise dir's contents as defined in the track repo with and student-submitted files added/overwriting.

What I would thus suggest is to add the `Ballerina.toml` and `Dependencies.toml` files to all the tests in the `tests/*` directory. That makes it a more representative situation.

File: bin/run-tests-in-docker.sh [hunk 8]
```diff
@@ -15,17 +15,19 @@
 # Stop executing when a command returns a non-zero return code
 set -e
 
+# Add a ballerina sample to pull dependencies
+mkdir -p tmp/ballerina
+cp -r tests/example-success/* tmp/ballerina
```
Comment (ErikSchierboom) [inline bin/run-tests-in-docker.sh:20]: This shouldn't be part of this script but done in the Dockerfile, as otherwise the deploy will fail (which doesn't call this script but just calls `docker build`). I'll do a follow-up PR after merging this PR.

**Review by ErikSchierboom — APPROVED** [review]
Awesome work! I've left some comments, but none of them should prevent this PR from being merged.

Side note: I've opened another PR to make the test runner available in production: https://github.com/exercism/terraform/pull/83/files Once that is merged and our terraform deploy has run, we can test things a bit on production.

**Comment by ErikSchierboom** [comment]
Merged! 🎉

**Comment by kytrinyx** [comment]
@vordimous in case you're not seeing pings in the ballerina repo, would you take a look at this one? https://github.com/exercism/ballerina/issues/146

File: bin/run-in-docker.sh (hunk 1 above)
Comment (vordimous) [reply bin/run-in-docker.sh]: I tried that and still got permission-denied errors. It could be an issue with the base `ballerina` image, but I haven't gone down that rabbit hole yet.

File: bin/run.sh (hunk 7 above)
Comment (vordimous) [reply bin/run.sh:46]: Ok, that makes it easier. I will update to assume these exist. The build command will report issues with these files.

File: bin/run-tests-in-docker.sh (hunk 8 above)
Comment (vordimous) [reply bin/run-tests-in-docker.sh:20]: So this is due to the need for a common place to copy the ballerina project to cache the correct libraries. I will comment on the other PR.
